### ISSUE TITLE
getOptimalNewCameraMatrix returns unstable results in opencv versions after 4.7.0

### ISSUE THREAD

**Issue by cirvladimir** [opencv/opencv#24923]
### System Information

OpenCV python version: 4.9.0
Operating System / Platform: Ubuntu 22.04
Python version: 3.10.12

### Detailed description

I noticed that getOptimalNewCameraMatrix returns unstable results for some very specific input matrices and distortions. If I change the input C_x by 0.1, the returned matrix C_x jumps by 80.
I also noticed that the unstable results happen on opencv version 4.8.2, 4.9.0, but do not happen on version 4.7.0. I'm wondering if this is a regression, or if I'm just unlucky with the numbers I have.
I did test with 50+ other very similar camera matrices and distortions, and only one of them had this instability. The unstable one (shown below) was within 2 sigmas of the median of the other matrices/distortion coefficients.

Run the python code below to reproduce this issue. On the stable version, Cx and Cy are near the center of the image. In the unstable versions, I get Cx and Cy at 500, 480, which is way off. Also in the unstable version, changing Cx from 632.11602783 to 632.0 changes the resulting Cx by 90.

### Steps to reproduce

1. Run the code below
2. Run the code below, but change 632.11602783 to 632.0. The resulting matrix changes by much more than it is supposed to change.

```python
import cv2
import numpy as np

cv2.getOptimalNewCameraMatrix(
    np.array(
        [
            [571.87939453, 0.0, 632.11602783],
            [0.0, 571.36779785, 405.00024414],
            [0.0, 0.0, 1.0],
        ]
    ),
    np.array(
        [
            2.95129061e-01, -4.90640551e-02, -1.27379884e-04, -2.73059668e-05, -1.65160082e-03, 6.26023889e-01, -3.09929363e-02, -1.06598334e-02, 0.00000000e00, 0.00000000e00, 0.00000000e00, 0.00000000e00, 0.00000000e00, 0.00000000e00,
        ]
    ),
    (1280, 800),
    1,
    (1280, 800),
)

```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by JStech** [comment]
The `-1`s added on lines 2270 and 2271 of `calibration.cpp` in [this commit](https://github.com/opencv/opencv/commit/7ea6b356c782ae8ae76e6bb5c49f6a77ef5bbf7e) are what cause the large difference. This function undistorts an array of points to detect what impact is has on the border of the image, and the `-1` is adjusting where exactly the border is. Maybe the distortion is large enough at the edge of the image that the one pixel difference has a big impact? I'd have to dig into this more to understand which behavior is correct.

**Comment by JStech** [comment]
To get a sense for what's going on, I used these distortion parameters to undistort a checkerboard:
![undistorted](https://github.com/opencv/opencv/assets/519268/08ff5aef-4ad4-4938-ae51-70e5e03c09fd)

Obviously, these are pretty severe parameters--it looks like the radial distortion even goes through a singularity. I think my prior guess was right, and the small change of subtracting 1 from the image dimensions gets magnified by the severe distortion.

With so much distortion, you should consider using [cv.fisheye](https://docs.opencv.org/3.4/db/d58/group__calib3d__fisheye.html).

**Comment by cirvladimir** [comment]
Thanks for looking into this.

I am using wide-angle lenses with about 120 degree horizontal FOV. I thought fisheye was for close to 180 degrees or more, so I used the non-fisheye model.
I think I got pretty decent reprojection errors using this model on the part of the image I care about.
Almost all lenses I calibrated have a singularity at some point at the corners of the image.
I though that was normal and it hasn't caused issues (except for this one specific matrix + distortion combo).

But it makes sense that I'm using this camera model at it's limits, so that's why I'm seeing this weird behavior. I guess we can close the issue.
